types/utils.go

```diff
@@ -95,6 +99,27 @@ func NewLevelDB(name, dir string) (db dbm.DB, err error) {
 	return dbm.NewDB(name, backend, dir)
 }
 
+// NewDB instantiate a new DB instance.
+// This differs from the tendermint/tm-db.NewDB function in three ways:
+// 1) Returns an error instead of panicking.
+// 2) Takes in a string for the backend type.
+// 3) If the backendType is an empty string, "goleveldb" is used.
+func NewDB(name, backendType, dir string) (db dbm.DB, err error) {
```

Comment:
Where does `err` actually get returned?

Reply:
I would strongly prefer that `dbm.NewDB` returns error rather than panic. This way we don't need to do this ugly recovery. 
But that will be a new breaking change. Do you think we can do it? I'm in support for that.

Reply:
Strong +1 to @robert-zaremba here. `panics` are not alternative mechanisms for error handling. Constructors like NewDB should _definitely_ not `recover`.

Reply:
The `dbm.NewDB` function comes from the Tendermint library:
```golang
dbm "github.com/tendermint/tm-db"
```

It returns the same things as this new `types.NewDB` method as (well as the deprecated `types.NewLevelDB`):
```golang
func NewDB(name string, backend BackendType, dir string) (DB, error) {
```

The `panic` recovery is in this new `types.NewDB` function because it was in the previous `types.NewLevelDB` function. All that `types.NewLevelDB` did was a) call `dbm.NewDB` with the type from `types.DBBackend`, and b) recover from a panic.

Looks like there was a conversation similar to this back when it was created: https://github.com/cosmos/cosmos-sdk/pull/3792/files#r262019126. At the time, though, `dbm.NewDB` didn't return an error. At some point, that was changed to return an error, and `types.NewLevelDB` was updated, but the `panic` recovery was left in: https://github.com/cosmos/cosmos-sdk/pull/6892/files#diff-d01ffec653bd293515fc86b07f11297dba8e9ac3764a32960ca4e6aee0f5233a

If you all are comfortable with removing the panic recovery, I'll remove it, but my instinct says it's worthwhile. It does return an `error`, but there's usually so much stuff involved with db stuff, that I don't usually trust that all possible error states have been accounted for.

Reply:
I removed the panic recovery, `defer`, and named return variables from `types.NewDB`.

I'm now tempted to get rid of the `types.NewDB` function altogether (and just say to use the `dbm.NewDB` function). All this `types.NewDB` does is:
1. Define a default of `goleveldb`. The only time this default is used is when the `config.toml` (Tendermint config file) hasn't been loaded yet.
2. Take in the type as a string instead of `BackendType` (which is a `string` type anyway). It might be nice to use the named variables for the hard-coded entries anyway.
3. Add a little extra context to any error messages.
4. Makes it clearer which DBs are specific to the Cosmos-SDK (as opposed to the Tendermint ones). This isn't much of a thing, though since `dbm.NewDB` is only used by. `types.NewDB` and the soon-to-be-deprecated `types.NewLevelDB`.

Thoughts?